src/main/java/com/google/devtools/build/lib/skyframe/WatchmanDiffAwareness.java

```diff
@@ -0,0 +1,247 @@
+    // query.
+    var ignored = collectChangesAndUpdateClockspec(eventHandler);
+  }
+
+  private ImmutableSet<Path> collectChangesAndUpdateClockspec(EventHandler eventHandler)
+      throws IOException, BrokenDiffAwarenessException {
+    sendCommand(nextQueryCommand);
+
+    boolean isFreshInstance = false;
+    ImmutableSet.Builder<Path> modifiedPaths = ImmutableSet.builder();
+    var json = new JsonReader(new BufferedReader(Channels.newReader(watchmanChannel, UTF_8)));
+    json.beginObject();
+    while (json.hasNext()) {
+      switch (json.nextName()) {
+        case "error" -> throw new BrokenDiffAwarenessException(json.nextString());
+        case "warning" -> {
+          String warning = json.nextString();
+          logger.atInfo().log("Watchman warning: %s", warning);
+          eventHandler.handle(
+              Event.warn(
+                  "Watchman returned a warning while checking for changed files: " + warning));
+        }
+        case "clock" -> updateClockspec(json.nextString());
+        case "is_fresh_instance" -> isFreshInstance = json.nextBoolean();
+        case "files" -> {
+          json.beginArray();
+          while (json.hasNext()) {
+            modifiedPaths.add(watchRootPath.resolve(json.nextString()));
+          }
+          json.endArray();
+        }
+        default -> json.skipValue();
+      }
+    }
+    json.endObject();
+
+    if (isFreshInstance) {
+      return null;
+    }
+    return modifiedPaths.build();
+  }
+
+  private List<Object> makeInitialQueryCommand(String watchRoot, Optional<String> watchPrefix) {
+    Map<String, Object> queryParams = new LinkedHashMap<>();
+    queryParams.put("fields", List.of("name"));
+    // Avoid an unnecessarily long response on the first query by omitting the list of potentially
+    // changed (thus at that point, all) files.
+    queryParams.put("empty_on_fresh_instance", true);
+    watchPrefix.ifPresent(prefix -> queryParams.put("relative_root", prefix));
+    // Bazel only needs to know about changed files, not directories.
+    // Watchman doesn't follow symlinks and treats junctions as symlinks, so there's no need to
+    // filter out the convenience symlinks.
+    // https://github.com/facebook/watchman/blob/13d4454123e6321408ba1610e4caa4f49863409f/watchman/fs/FileInformation.cpp#L70
+    List<Object> anyof =
+        Stream.concat(
+                Stream.of("anyof", List.of("type", "d")),
+                ignoredPaths.stream().map(ignoredPath -> List.of("dirname", ignoredPath)))
+            .toList();
+    queryParams.put("expression", List.of("not", anyof));
+    return List.of("query", watchRoot, queryParams);
+  }
+
+  private void updateClockspec(String clockspec) {
+    // The query command has the form ["query", "/foo/bar", {"since": ..., ...}].
+    ((Map<String, Object>) nextQueryCommand.get(2)).put("since", clockspec);
+  }
+
+  private void sendCommand(List<Object> command) throws IOException {
```

Comment:
Nit: It seems like we are using sendCommand with 2 queries right now:

- "watch-project" with path to bazel workspace

- "query" for file changes.

For both of these use cases, the command contents are relatively static enough that plain string templating could be easier/faster than serializing List<Object> to string with GSON. WDYT?

Reply:
Yes, I'm considering using GSON for the initial JSON -> string conversion with a placeholder string for the `since` argument that is then replaced via pure string operations for all future commands. I would like to keep this simple for the initial review and benchmark it properly before performing such optimizations tough.

Reply:
Some rough profiling results for switching from `master` to `release-7.2.0` with roughly 3,000 changed files: `collectChangesAndUpdateClockspec` takes 20ms and `sendCommand` accounts for only 0.1ms of that, so I think we can safely ignore it.